src/Compilers/CSharp/Portable/Binder/SwitchBinder_BindPatternSwitch.cs

```diff
@@ -63,67 +198,402 @@ private BoundPatternSwitchSection BindPatternSwitchSection(BoundExpression bound
+                        if (!hasErrors && pattern.ConstantValue != null && this.FindMatchingSwitchCaseLabel(pattern.ConstantValue) != label)
+                        {
+                            diagnostics.Add(ErrorCode.ERR_DuplicateCaseLabel, node.Location, pattern.ConstantValue?.Value ?? label.Name);
+                            hasErrors = true;
+                        }
+                        return new BoundPatternSwitchLabel(node, label, pattern, null, hasErrors);
                     }
 
                 case SyntaxKind.DefaultSwitchLabel:
                     {
                         var defaultLabelSyntax = (DefaultSwitchLabelSyntax)node;
-                        var pattern = new BoundWildcardPattern(node);
+                        var pattern = new BoundWildcardPattern(node, knownMatchResult: true);
+                        bool hasErrors = pattern.HasErrors;
                         if (defaultLabel != null)
                         {
                             diagnostics.Add(ErrorCode.ERR_DuplicateCaseLabel, node.Location, "default");
+                            hasErrors = true;
                         }
 
                         defaultLabel = defaultLabelSyntax;
-                        return new BoundPatternSwitchLabel(node, pattern, null, node.HasErrors); // note that this is semantically last!
+                        // Note that this is semantically last! The caller will place it in the decision tree
+                        // in the final position.
+                        return new BoundPatternSwitchLabel(node, label, pattern, null, hasErrors);
+                    }
+
+                case SyntaxKind.CasePatternSwitchLabel:
+                    {
+                        var matchLabelSyntax = (CasePatternSwitchLabelSyntax)node;
+                        var pattern = sectionBinder.BindPattern(
+                            matchLabelSyntax.Pattern, boundSwitchExpression, switchExpressionIsNull, boundSwitchExpression.Type, node.HasErrors, diagnostics, wasSwitchCase: true);
+                        return new BoundPatternSwitchLabel(node, label, pattern,
+                            matchLabelSyntax.WhenClause != null ? sectionBinder.BindBooleanExpression(matchLabelSyntax.WhenClause.Condition, diagnostics) : null, node.HasErrors);
                     }
 
                 default:
-                    throw ExceptionUtilities.Unreachable;
+                    throw ExceptionUtilities.UnexpectedValue(node);
+            }
+        }
+
+        private Dictionary<SyntaxNode, LabelSymbol> _labelsByNode;
+        private Dictionary<SyntaxNode, LabelSymbol> LabelsByNode
+        {
+            get
+            {
+                if (_labelsByNode == null)
+                {
+                    var result = new Dictionary<SyntaxNode, LabelSymbol>();
+                    foreach (var label in Labels)
+                    {
+                        var sl = label as SourceLabelSymbol;
```

Comment:
Please do not use ```as``` cast if you don't expect it to fail. Otherwise, the code should handle the null result properly.